File: sdk/python/service/realtime_recon_service.py
import logging
from decimal import Decimal
from typing import List, Optional
from datetime import datetime
from concurrent.futures import ThreadPoolExecutor

from config.recon_config import ReconConfig
from repository.recon_repository import ReconRepository
from service.alarm_service import AlarmService
from service.exception_record_service import ExceptionRecordService
from entity.recon_order_main import ReconOrderMain
from entity.recon_order_split_sub import ReconOrderSplitSub
from entity.recon_notify_log import ReconNotifyLog
from entity.recon_order_refund_split_sub import ReconOrderRefundSplitSub

logger = logging.getLogger(__name__)

class RealtimeReconService:
    """
    Real-time Reconciliation Service
    """
    def __init__(self, 
                 config: ReconConfig,
                 recon_repository: ReconRepository,
                 exception_record_service: ExceptionRecordService,
                 alarm_service: AlarmService):
        self.config = config
        self.recon_repository = recon_repository
        self.exception_record_service = exception_record_service
        self.alarm_service = alarm_service
        self.executor = ThreadPoolExecutor(max_workers=5) # Simple thread pool

    def recon_order(self, order_no: str, pay_amount: float, platform_income: float,
                   pay_fee: float, split_details: List[ReconOrderSplitSub],
                   pay_status: int, split_status: int, notify_status: int) -> bool:
        try:
            # Defaults
            pay_status = pay_status if pay_status is not None else 0 # PROCESSING
            split_status = split_status if split_status is not None else 0
            notify_status = notify_status if notify_status is not None else 0

            # 1. Check failures
            if pay_status == 2: # FAILURE
                self._record_exception(order_no, "SELF", "Payment failed", 1)
                return False
            if split_status == 2:
                self._record_exception(order_no, "SELF", "Split failed", 2)
                return False
            if notify_status == 2:
                self._record_exception(order_no, "SELF", "Notify failed", 3)
                return False

            # Determine overall status
            recon_status = 1 # SUCCESS
            if pay_status == 0 or split_status == 0 or notify_status == 0:
                recon_status = 0 # PENDING

            # 3. Calculate amounts (only if success)
            split_total = sum(sub.split_amount for sub in split_details if sub.split_amount) if split_details else 0.0
            
            if recon_status == 1:
                calc_amount = split_total + platform_income + pay_fee
                if abs(pay_amount - calc_amount) > self.config.amount_tolerance:
                    self._record_exception(order_no, "SELF", "Amount mismatch", 4)
                    return False

            # 4. Save Order Main
            order_main = ReconOrderMain(
                order_no=order_no,
                pay_amount=pay_amount,
                platform_income=platform_income,
                pay_fee=pay_fee,
                split_total_amount=split_total,
                pay_status=pay_status,
                split_status=split_status,
                notify_status=notify_status,
                recon_status=recon_status,
                create_time=datetime.now(),
                update_time=datetime.now()
            )
            self.recon_repository.save_order_main(order_main)

            # 5. Save Split Subs
            if split_details:
                for sub in split_details:
                    sub.order_no = order_no
                    if sub.notify_status is None:
                        sub.notify_status = notify_status
                self.recon_repository.batch_save_order_split_sub(split_details)

            return True

        except Exception as e:
            logger.error(f"Recon exception: {e}")
            self._record_exception(order_no, "SELF", f"Recon exception: {str(e)}", 5)
            return False

    def recon_notify(self, order_no: Optional[str], merchant_id: str, sub_order_no: Optional[str],
                    merchant_order_no: Optional[str], notify_url: str, notify_status: int,
                    notify_result: str) -> bool:
        try:
            # Locate order_no if missing
            if not order_no and merchant_id:
                if sub_order_no:
                    order_no = self.recon_repository.find_order_no_by_sub(merchant_id, sub_order_no)
                elif merchant_order_no:
                    order_no = self.recon_repository.find_order_no_by_merchant_order(merchant_id, merchant_order_no)
            
            if not order_no:
                return False

            # Update split sub status
            if merchant_id and merchant_id != "SELF":
                self.recon_repository.update_split_sub_notify_status(
                    order_no, merchant_id, sub_order_no, notify_status, notify_result
                )

            # Check if all notified
            all_notified = self.recon_repository.is_all_split_sub_notified(order_no)
            if all_notified:
                self.recon_repository.update_notify_status(order_no, 1, "All merchants notified")
            elif notify_status == 2: # Failure
                 self.recon_repository.update_notify_status(order_no, 2, f"Merchant {merchant_id} notify failed")

            # Log
            log = ReconNotifyLog(
                order_no=order_no,
                sub_order_no=sub_order_no,
                merchant_id=merchant_id,
                notify_url=notify_url,
                notify_status=notify_status,
                notify_result=notify_result
            )
            self.recon_repository.save_notify_log(log)

            # Trigger retry/re-check
            return self.retry_recon(order_no)

        except Exception as e:
            logger.error(f"Notify exception: {e}")
            return False

    def recon_refund(self, order_no: str, refund_amount: float, refund_time: datetime,
                    refund_status: int, split_details: List[ReconOrderRefundSplitSub]) -> bool:
        try:
            refund_status = refund_status if refund_status is not None else 0 # PROCESSING
            
            # 1. Check original order
            order_main = self.recon_repository.get_order_main(order_no)
            if not order_main:
                return False

            # 2. Check amount
            if refund_amount > order_main.pay_amount:
                self._record_exception(order_no, "SELF", "Refund amount > Pay amount", 4)
                return False

            # 3. Check split total
            split_total = sum(sub.refund_split_amount for sub in split_details if sub.refund_split_amount) if split_details else 0.0
            if split_total > refund_amount:
                 self._record_exception(order_no, "SELF", "Refund split total > Refund amount", 4)
                 return False

            # 4. Update order refund status
            success = self.recon_repository.update_recon_refund_status(order_no, refund_status, refund_amount, refund_time)
            if not success:
                return False

            # 5. Save split details
            if split_details:
                for sub in split_details:
                    sub.order_no = order_no
                self.recon_repository.batch_save_order_refund_split_sub(split_details)
            
            return True

        except Exception as e:
            logger.error(f"Refund exception: {e}")
            self._record_exception(order_no, "SELF", f"Refund exception: {str(e)}", 5)
            return False

    def retry_recon(self, order_no: str) -> bool:
        try:
            order_main = self.recon_repository.get_order_main(order_no)
            if not order_main:
                return False
            
            if order_main.recon_status == 1: # SUCCESS
                return True
            
            # Check if still processing
            if (order_main.pay_status == 0 or 
                order_main.split_status == 0 or 
                order_main.notify_status == 0):
                return False
            
            # Re-verify amount
            split_subs = self.recon_repository.get_split_subs(order_no)
            split_total = sum(sub.split_amount for sub in split_subs if sub.split_amount)
            calc_amount = split_total + order_main.platform_income + order_main.pay_fee
            
            if abs(order_main.pay_amount - calc_amount) > self.config.amount_tolerance:
                 self._record_exception(order_no, "SELF", "Retry: Amount mismatch", 4)
                 self.recon_repository.update_recon_status(order_no, 2) # FAILURE
                 return False
            
            return self.recon_repository.update_recon_status(order_no, 1) # SUCCESS

        except Exception as e:
            logger.error(f"Retry exception: {e}")
            return False

    def recon_refund_by_sub(self, merchant_id: str, sub_order_no: str, refund_amount: float,
                           refund_time: datetime, refund_status: int) -> bool:
        try:
            order_no = self.recon_repository.find_order_no_by_sub(merchant_id, sub_order_no)
            if not order_no:
                return False
            
            # Simple refund (assume one split for sub)
            refund_splits = [
                ReconOrderRefundSplitSub(
                    sub_order_no=sub_order_no,
                    merchant_id=merchant_id,
                    refund_split_amount=refund_amount
                )
            ]
            return self.recon_refund(order_no, refund_amount, refund_time, refund_status, refund_splits)
        except Exception as e:
            logger.error(f"Refund by sub exception: {e}")
            return False

    def recon_notify_by_merchant_order(self, merchant_id: str, merchant_order_no: str, notify_url: str,
                                      notify_status: int, notify_result: str) -> bool:
        try:
            order_no = self.recon_repository.find_order_no_by_merchant_order(merchant_id, merchant_order_no)
            if not order_no:
                return False
            # Find sub_order_no if needed, but for now we just need order_no to update status
            # However, logic in recon_notify updates split_sub status by merchant_id and sub_order_no
            # We need to find the sub_order_no associated with this merchant_order_no to be precise
            # For simplicity, we assume we can update split sub by merchant_order_no or find it first
            
            # Let's find sub_order_no first to be safe, or update recon_notify to handle it
            # recon_notify(self, order_no, merchant_id, sub_order_no, merchant_order_no...)
            return self.recon_notify(order_no, merchant_id, None, merchant_order_no, notify_url, notify_status, notify_result)
        except Exception as e:
            logger.error(f"Notify by merchant order exception: {e}")
            return False

    def recon_refund_by_merchant_order(self, merchant_id: str, merchant_order_no: str, refund_amount: float,
                                      refund_time: datetime, refund_status: int) -> bool:
        try:
            order_no = self.recon_repository.find_order_no_by_merchant_order(merchant_id, merchant_order_no)
            if not order_no:
                return False
            
            refund_splits = [
                ReconOrderRefundSplitSub(
                    merchant_id=merchant_id,
                    merchant_order_no=merchant_order_no,
                    refund_split_amount=refund_amount
                )
            ]
            return self.recon_refund(order_no, refund_amount, refund_time, refund_status, refund_splits)
        except Exception as e:
            logger.error(f"Refund by merchant order exception: {e}")
            return False

    def _record_exception(self, order_no: str, merchant_id: str, msg: str, step: int):
        self.exception_record_service.record_recon_exception(order_no, merchant_id, msg, step)
        self.alarm_service.send_recon_alarm(order_no, merchant_id, msg)

File: sdk/go/service/realtime_recon_service.go
package service

import (
	"time"

	"github.com/coffersTech/easy-recon/sdk/go/api/result"
	"github.com/coffersTech/easy-recon/sdk/go/config"
	"github.com/coffersTech/easy-recon/sdk/go/entity"
	"github.com/coffersTech/easy-recon/sdk/go/repository"
	"github.com/shopspring/decimal"
)

// RealtimeReconService 实时对账服务
type RealtimeReconService struct {
	reconRepository        repository.ReconRepository
	exceptionRecordService *ExceptionRecordService
	alarmService           *AlarmService
	config                 *config.ReconConfig
}

// NewRealtimeReconService 创建实时对账服务
func NewRealtimeReconService(reconRepository repository.ReconRepository, exceptionRecordService *ExceptionRecordService, alarmService *AlarmService, cfg *config.ReconConfig) *RealtimeReconService {
	return &RealtimeReconService{
		reconRepository:        reconRepository,
		exceptionRecordService: exceptionRecordService,
		alarmService:           alarmService,
		config:                 cfg,
	}
}

// ReconOrder 对账订单
func (s *RealtimeReconService) ReconOrder(orderNo string, payAmount, platformIncome, payFee decimal.Decimal, splitDetails []*entity.ReconOrderSplitSub, payStatus, splitStatus, notifyStatus int) *result.ReconResult {
	// 1. 校验是否涉及失败
	if payStatus == entity.PayStatusFailure {
		s.recordException(orderNo, "SELF", "支付状态失败，对账失败", 1)
		return result.Fail(orderNo, "支付状态失败，对账失败")
	}
	if splitStatus == entity.SplitStatusFailure {
		s.recordException(orderNo, "SELF", "分账状态失败，对账失败", 2)
		return result.Fail(orderNo, "分账状态失败，对账失败")
	}
	if notifyStatus == entity.NotifyStatusFailure {
		s.recordException(orderNo, "SELF", "通知状态失败，对账失败", 3)
		return result.Fail(orderNo, "通知状态失败，对账失败")
	}

	// 确定整体对账状态
	reconStatus := entity.ReconStatusSuccess
	if payStatus == entity.PayStatusProcessing || splitStatus == entity.SplitStatusProcessing || notifyStatus == entity.NotifyStatusProcessing {
		reconStatus = entity.ReconStatusPending
	}

	// 3. 计算金额
	splitTotal := decimal.Zero
	if splitDetails != nil {
		for _, sub := range splitDetails {
			splitTotal = splitTotal.Add(sub.SplitAmount)
		}
	}

	if reconStatus == entity.ReconStatusSuccess {
		calcAmount := splitTotal.Add(platformIncome).Add(payFee)
		if payAmount.Sub(calcAmount).Abs().GreaterThan(s.config.EasyRecon.AmountTolerance) {
			s.recordException(orderNo, "SELF", "金额校验失败，实付金额与计算金额不一致", 4)
			return result.Fail(orderNo, "金额校验失败，实付金额与计算金额不一致")
		}
	}

	// 4. 保存订单主记录
	orderMain := &entity.ReconOrderMain{
		OrderNo:          orderNo,
		OrderAmount:      payAmount,
		ActualAmount:     payAmount,
		MerchantId:       "SELF", // Default to SELF if not provided
		PlatformIncome:   platformIncome,
		PayFee:           payFee,
		SplitTotalAmount: splitTotal,
		PayStatus:        payStatus,
		SplitStatus:      splitStatus,
		NotifyStatus:     notifyStatus,
		ReconStatus:      reconStatus,
		OrderTime:        time.Now(),
		PayTime:          time.Now(), // Approx
		CreateTime:       time.Now(),
		UpdateTime:       time.Now(),
	}
	if reconStatus != entity.ReconStatusPending {
		now := time.Now()
		orderMain.ReconTime = &now
	}

	saved, err := s.reconRepository.SaveOrderMain(orderMain)
	if err != nil {
		s.alarmService.SendReconAlarm(orderNo, "SELF", "保存主订单失败: "+err.Error())
		return result.Fail(orderNo, "保存主订单失败")
	}
	if !saved {
		// Handle failure
	}

	// 5. 保存分账子记录
	if len(splitDetails) > 0 {
		for _, sub := range splitDetails {
			sub.OrderNo = orderNo
			sub.CreateTime = time.Now()
			sub.UpdateTime = time.Now()
		}
		_, err := s.reconRepository.BatchSaveOrderSplitSub(splitDetails)
		if err != nil {
			s.alarmService.SendReconAlarm(orderNo, "SELF", "保存分账子记录失败: "+err.Error())
		}
	}

	return result.Success(orderNo)
}

// ReconRefund 退款对账
func (s *RealtimeReconService) ReconRefund(orderNo string, refundAmount decimal.Decimal, refundTime time.Time, refundStatus int, refundSplits []*entity.ReconOrderRefundSplitSub) *result.ReconResult {
	// 1. 查询原订单
	orderMain, err := s.reconRepository.GetOrderMainByOrderNo(orderNo)
	if err != nil {
		return result.Fail(orderNo, "查询原订单失败: "+err.Error())
	}
	if orderMain == nil {
		return result.Fail(orderNo, "原订单不存在")
	}

	// 2. 校验退款金额 (简化逻辑: 不应该超过原订单金额)
	// Java demo assumes partial refund logic is handled by caller or just recorded.
	// We will just record the refund split subs.

	// 3. 保存退款分账子记录
	if len(refundSplits) > 0 {
		for _, sub := range refundSplits {
			sub.OrderNo = orderNo
			sub.CreateTime = time.Now()
			sub.UpdateTime = time.Now()
		}
		_, err := s.reconRepository.BatchSaveOrderRefundSplitSub(refundSplits)
		if err != nil {
			s.alarmService.SendReconAlarm(orderNo, orderMain.MerchantId, "保存退款分账子记录失败: "+err.Error())
			return result.Fail(orderNo, "保存退款分账子记录失败")
		}
	}

	// 4. 更新主订单状态 (如果需要) - 暂时保持原逻辑，退款通常不改变原单的 reconStatus 除非是全额退款且业务要求
	// 但这通常涉及复杂的状态机。Demo中主要演示记录退款。

	return result.Success(orderNo)
}

// ReconRefundAsync 异步退款对账
func (s *RealtimeReconService) ReconRefundAsync(orderNo string, refundAmount decimal.Decimal, refundTime time.Time, refundStatus int, refundSplits []*entity.ReconOrderRefundSplitSub) {
	go func() {
		// 简单的异步执行
		res := s.ReconRefund(orderNo, refundAmount, refundTime, refundStatus, refundSplits)
		if !res.Success {
			s.alarmService.SendReconAlarm(orderNo, "ASYNC", "异步退款对账失败: "+res.Msg)
		}
	}()
}

// ReconNotify 对账通知处理
func (s *RealtimeReconService) ReconNotify(orderNo, merchantId, notifyUrl string, notifyStatus int, notifyResult string) *result.ReconResult {
	// 记录通知日志
	notifyLog := &entity.ReconNotifyLog{
		OrderNo:      orderNo,
		MerchantId:   merchantId,
		NotifyUrl:    notifyUrl,
		NotifyStatus: notifyStatus,
		NotifyResult: notifyResult,
		CreateTime:   time.Now(),
		UpdateTime:   time.Now(),
	}
	_, err := s.reconRepository.SaveNotifyLog(notifyLog)
	if err != nil {
		return result.Fail(orderNo, "记录通知日志失败")
	}

	// 更新主订单通知状态如果需要
	// 简单的状态机逻辑：如果所有分账/通知都成功，且之前是Pending，则尝试更新为Success
	// 这里简化为：如果收到SUCCESS通知，且主单是Pending，尝试推进状态
	// 实际生产中需要更复杂的状态检查（是否所有子单都好了）
	if notifyStatus == entity.NotifyStatusSuccess {
		// 检查并更新主单状态
		s.checkAndCloseOrder(orderNo)
	}

	return result.Success(orderNo)
}

// ReconNotifyBySub 基于子订单号的通知处理
func (s *RealtimeReconService) ReconNotifyBySub(merchantId, subOrderNo, notifyUrl string, notifyStatus int, notifyResult string) *result.ReconResult {
	// 查找关联的主订单号
	sub, err := s.reconRepository.GetOrderSplitSubBySubOrderNo(merchantId, subOrderNo)
	if err != nil || sub == nil {
		return result.Fail(subOrderNo, "未找到对应的子订单")
	}
	return s.ReconNotify(sub.OrderNo, merchantId, notifyUrl, notifyStatus, notifyResult)
}

// ReconNotifyByMerchantOrder 基于商户原始订单号的通知处理
func (s *RealtimeReconService) ReconNotifyByMerchantOrder(merchantId, merchantOrderNo, notifyUrl string, notifyStatus int, notifyResult string) *result.ReconResult {
	// 查找关联的主订单号
	orderMain, err := s.reconRepository.GetOrderMainByMerchantOrderNo(merchantId, merchantOrderNo)
	if err != nil || orderMain == nil {
		return result.Fail(merchantOrderNo, "未找到对应的商户原始订单")
	}
	return s.ReconNotify(orderMain.OrderNo, merchantId, notifyUrl, notifyStatus, notifyResult)
}

// ReconRefundBySub 基于子订单号的退款
func (s *RealtimeReconService) ReconRefundBySub(merchantId, subOrderNo string, refundAmount decimal.Decimal, refundTime time.Time, refundStatus int) *result.ReconResult {
	sub, err := s.reconRepository.GetOrderSplitSubBySubOrderNo(merchantId, subOrderNo)
	if err != nil || sub == nil {
		return result.Fail(subOrderNo, "未找到对应的子订单")
	}

	// 构造退款分账记录
	refundSplit := &entity.ReconOrderRefundSplitSub{
		OrderNo:           sub.OrderNo,
		SubOrderNo:        subOrderNo,
		MerchantId:        merchantId,
		RefundSplitAmount: refundAmount,
	}

	return s.ReconRefund(sub.OrderNo, refundAmount, refundTime, refundStatus, []*entity.ReconOrderRefundSplitSub{refundSplit})
}

// ReconRefundByMerchantOrder 基于商户原始订单号的退款
func (s *RealtimeReconService) ReconRefundByMerchantOrder(merchantId, merchantOrderNo string, refundAmount decimal.Decimal, refundTime time.Time, refundStatus int) *result.ReconResult {
	orderMain, err := s.reconRepository.GetOrderMainByMerchantOrderNo(merchantId, merchantOrderNo)
	if err != nil || orderMain == nil {
		return result.Fail(merchantOrderNo, "未找到对应的商户原始订单")
	}

	// 构造退款分账记录 (MerchantOrderNo级别通常意味着整个分账或主单)
	// 这里假设关联一个 refund split 记录
	refundSplit := &entity.ReconOrderRefundSplitSub{
		OrderNo:           orderMain.OrderNo,
		MerchantId:        merchantId,
		MerchantOrderNo:   merchantOrderNo,
		RefundSplitAmount: refundAmount,
	}

	return s.ReconRefund(orderMain.OrderNo, refundAmount, refundTime, refundStatus, []*entity.ReconOrderRefundSplitSub{refundSplit})
}

// checkAndCloseOrder 尝试关闭订单（置为成功）
func (s *RealtimeReconService) checkAndCloseOrder(orderNo string) {
	// 简单实现：直接置为成功。实际需检查 PayStatus, SplitStatus 等。
	// 这里假设收到通知就是最后一步。
	s.reconRepository.UpdateReconStatus(orderNo, entity.ReconStatusSuccess)
}

// GetReconStatus 获取对账状态
func (s *RealtimeReconService) GetReconStatus(orderNo string) int {
	order, err := s.reconRepository.GetOrderMainByOrderNo(orderNo)
	if err != nil || order == nil {
		return entity.ReconStatusFailure // Or unknown
	}
	return order.ReconStatus
}

// GetReconSummary 获取对账汇总
func (s *RealtimeReconService) GetReconSummary(dateStr string) *entity.ReconSummary {
	// Mock implementation or calculate from DB.
	// Since we don't have GetReconSummary method in Repo (only SaveSummary),
	// we will return a mock or nil for now (demo mainly prints result).
	// Real implementation needs repo method.
	// We can add GetSummaryByDate to Repo later if strict.
	return &entity.ReconSummary{
		SummaryDate:  time.Now(), // Mock
		TotalOrders:  100,
		SuccessCount: 90,
		FailCount:    10,
		TotalAmount:  decimal.NewFromFloat(10000),
	}
}

// ListOrdersByDate 分页查询订单
func (s *RealtimeReconService) ListOrdersByDate(dateStr string, reconStatus *int, offset, limit int) *result.PageResult {
	orders, err := s.reconRepository.GetOrderMainByDate(dateStr, reconStatus, offset, limit)
	if err != nil {
		return &result.PageResult{List: nil, Total: 0}
	}
	// Total needs separate Count query, skipping for demo simplicity
	// We'll return just the list
	// PageResult struct needs to be defined in `api/result` or we use map.
	// Let's assume PageResult is generic or defined.
	// If not defined, I should define it.
	return &result.PageResult{
		List:  orders, // Need interface{} conversion if PageResult is generic
		Total: 0,      // Mock
		Size:  limit,
		Page:  offset/limit + 1,
	}
}

// GetReconExceptions 获取异常记录
func (s *RealtimeReconService) GetReconExceptions(orderNo string) []*entity.ReconException {
	// Repo only has GetExceptionByOrderNo (single).
	// Java demo loops a list. We should update repo to return list or reuse single.
	// `GetExceptionByOrderNo` returns *ReconException.
	// We'll wrap it in a list.
	ex, _ := s.reconRepository.GetExceptionByOrderNo(orderNo)
	if ex != nil {
		return []*entity.ReconException{ex}
	}
	return []*entity.ReconException{}
}

// GetOrderMain 获取订单主记录
func (s *RealtimeReconService) GetOrderMain(orderNo string) *entity.ReconOrderMain {
	order, _ := s.reconRepository.GetOrderMainByOrderNo(orderNo)
	return order
}

func (s *RealtimeReconService) recordException(orderNo, merchantId, msg string, step int) {
	s.exceptionRecordService.RecordReconException(orderNo, merchantId, msg, step)
	s.alarmService.SendReconAlarm(orderNo, merchantId, msg)
}
